apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/SerializationHelpers.cs

```diff
@@ -27,12 +30,43 @@ internal static class SerializationHelpers
         /// <summary>
         /// Checks whether <paramref name="type"/> refers to an anonymous type.
         /// </summary>
-        internal static bool IsAnonymousType(System.Type type) => type.GetTypeInfo().IsDefined(typeof(CompilerGeneratedAttribute));
+        internal static bool IsAnonymousType(BclType type) => type.GetTypeInfo().IsDefined(typeof(CompilerGeneratedAttribute));
 
         /// <summary>
         /// Checks whether <paramref name="type"/> refers to a type annotated with <see cref="FirestoreDataAttribute"/>.
         /// </summary>
-        internal static bool IsFirestoreAttributedType(System.Type type) => type.GetTypeInfo().IsDefined(typeof(FirestoreDataAttribute));
+        internal static bool IsFirestoreAttributedType(BclType type) => type.GetTypeInfo().IsDefined(typeof(FirestoreDataAttribute));
+
+        /// <summary>
+        /// Checks whether <paramref name="type"/> refers to a non-nullable type (so a null value is invalid).
+        /// </summary>
+        internal static bool IsNonNullableValueType(BclType type) => type.GetTypeInfo().IsValueType && Nullable.GetUnderlyingType(type) == null;
+
+        /// <summary>
+        /// If <paramref name="type"/> implements <see cref="IDictionary{TKey, TValue}"/> with TKey equal to string, returns true and sets
+        /// <paramref name="elementType"/> to TValue. Otherwise, returns false and sets <paramref name="elementType"/> to null.
+        /// </summary>
+        internal static bool IsStringDictionaryType(BclType type, out BclType elementType)
+        {
+            elementType = type.GetTypeInfo().GetInterfaces().Select(MapInterfaceToDictionaryValueTypeArgument).FirstOrDefault(t => t != null);
```

Comment:
Any( x => ...) ?

Reply:
oh nvm you return it via out

Reply:
Isnt this normally called a "TryGet"...?

Reply:
Yes, good call - I'd gone through a bunch of iterations, but hadn't taken it to its logical conclusion. Will fix up.